zipline/data/loader.py

```diff
@@ -273,26 +247,68 @@ def ensure_treasury_data(bm_symbol, first_date, last_date, now):
         re-download data that isn't available due to scheduling quirks or other
         failures.
 
-    We attempt to download data unless we already have data stored in the cache
-    for `module_name` whose first entry is before or on `first_date` and whose
-    last entry is on or after `last_date`.
+    We attempt to download data unless we already have data stored in source or
+    in the cache for `module_name` whose first entry is before or on
+    `first_date` and whose last entry is on or after `last_date`.
 
     If we perform a download and the cache criteria are not satisfied, we wait
     at least one hour before attempting a redownload.  This is determined by
     comparing the current time to the result of os.path.getmtime on the cache
     path.
     """
     loader_module, filename, source = INDEX_MAPPING.get(
-        bm_symbol, INDEX_MAPPING['^GSPC']
+        symbol, INDEX_MAPPING['^GSPC'],
     )
     first_date = max(first_date, loader_module.earliest_possible_date())
+
+    data = _load_cached_data(filename, first_date, last_date, now, 'treasury')
+    if data is not None:
+        return data
+
+    # If no cached data was found or it was missing any dates then download the
+    # necessary data.
+    logger.info('Downloading treasury data for {symbol!r}.', symbol=symbol)
+
+    try:
+        data = loader_module.get_treasury_data(first_date, last_date)
+        data.to_csv(get_data_filepath(filename))
+    except (OSError, IOError, HTTPError):
+        logger.exception('failed to cache treasury data')
+    if not has_data_for_dates(data, first_date, last_date):
+        logger.warn("Still don't have expected data after redownload!")
```

Comment:
If we're consolidating these helper methods, I wonder if we could remove even this remaining duplication. Not a blocker.